AppWidget/app/src/main/java/com/example/android/appwidget/ItemsCollectionAppWidget.kt

```diff
@@ -0,0 +1,185 @@
+ */
+class ItemsCollectionAppWidget : AppWidgetProvider() {
+
+    override fun onUpdate(
+        context: Context,
+        appWidgetManager: AppWidgetManager,
+        appWidgetIds: IntArray
+    ) {
+        val remoteViews = RemoteViews(context.packageName, R.layout.widget_items_collection)
+        if (BuildCompat.isAtLeastS()) {
+            val collectionItems =
+                ItemsCollectionRemoteViewsFactory.getRemoteCollectionItems(context)
+            remoteViews.setRemoteAdapter(R.id.items_list_view, collectionItems)
+        } else {
+            remoteViews.setRemoteAdapter(
+                R.id.items_list_view,
+                Intent(context, ItemsCollectionRemoteViewsService::class.java)
+            )
+        }
+        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews)
+    }
+
+    override fun onReceive(context: Context?, intent: Intent?) {
+        super.onReceive(context, intent)
+        if (BuildCompat.isAtLeastS() &&
+            intent?.extras?.getInt(REQUEST_CODE) == REQUEST_CODE_FROM_COLLECTION_WIDGET
+        ) {
+            intent.extras?.let {
+                // The checked state is stored in the Bundle in the Intent
+                val checked = it.getBoolean(
+                    RemoteViews.EXTRA_CHECKED,
+                    false
+                )
+                Toast.makeText(
+                    context,
+                    "ViewId : ${it.getInt(EXTRA_VIEW_ID)}'s checked status is now : $checked",
+                    Toast.LENGTH_SHORT
+                ).show()
+            }
+        }
+    }
+}
+
+class ItemsCollectionRemoteViewsService : RemoteViewsService() {
+
+    override fun onGetViewFactory(data: Intent?): RemoteViewsFactory {
+        return ItemsCollectionRemoteViewsFactory(applicationContext)
+    }
+}
+
+class ItemsCollectionRemoteViewsFactory(private val context: Context) :
+    RemoteViewsService.RemoteViewsFactory {
+
+    companion object {
+        val items = listOf(
+            R.layout.item_checkboxes,
+            R.layout.item_radio_buttons,
+            R.layout.item_switches,
+        )
+
+        const val REQUEST_CODE_FROM_COLLECTION_WIDGET = 2
+        const val EXTRA_VIEW_ID = "extra_view_id"
+        const val REQUEST_CODE = "request_code"
+
+        @RequiresApi(31)
+        fun getRemoteCollectionItems(context: Context): RemoteViews.RemoteCollectionItems {
+            val builder = RemoteViews.RemoteCollectionItems.Builder()
+            items.forEachIndexed { index, layoutId ->
+                builder.addItem(index.toLong(), getRemoteViewsForLayoutId(context, layoutId))
+            }
+            return builder.setHasStableIds(true).setViewTypeCount(items.count()).build()
+        }
+
+        internal fun getRemoteViewsForLayoutId(
+            context: Context,
+            @LayoutRes layoutId: Int
+        ): RemoteViews {
+            val remoteViews = RemoteViews(context.packageName, layoutId)
+            if (BuildCompat.isAtLeastS()) {
```

Comment:
I think this could benefit from an early return:

```
if (!BuildCompat.isAtLeastS()) {
  return remoteViews;
}
...
```

Reply:
Done.